### ISSUE TITLE
Several REST API requests to make building an explorer easier

### ISSUE THREAD

**Issue by carlcherry** [hashgraph/hedera-mirror-node#1620]
**Problem**
Based on how our explorer works our only way of getting the needed information has been to use direct postgres access. This normally means we can't upgrade easily due to schema changes. Our specific needs are these:

1. Getting information about the very latest record_file added including the index, consensus_start, consensus_end, etc.
2. A query to get all the transactions for a record_file (by index ideally)
3. For transactions we would like to have all the details required to fully understand what has transpired - this is especially true for transactions that have failed or those with record fees. We have been using the non_fee_transfer table values to build a better picture.
4. We store the raw data of a transaction in our postgres db - this info is not available in the /transactions API.
5. It seems that if a transaction fails there are no records stored on the non_fee_transfer table - so the original intent of the transfer is lost.

**Solution**
1. We would like to have an API to get the latest index. Something like /api/v1/blocks/latest returns information about the most recent record_file.
2. It would be nice to be able to pull all transactions based on the record_file index.
3. For transactions it would be nice to get the actual transfers amounts for each account plus the original non_fee_transfer amounts
4. Add a include_raw parameter to the /transactions route and return the raw bytes if available.
5. Always store original transfer info to the non_fee_transfer table.

**Additional Context**
I have been using direct postgres access and this creates a lot of challenges with respect to upgrades.

**Comment by Nana-EC** [comment]
Hey @calvinchengx ,

Thanks for the suggestions.
Some follow up questions here, as a better understanding of the needs and applications might help us better design our approach to helping resolve some of these issues.

1. Can you elaborate a bit more on your use cases from a record approach for requests #1 & #2. Is this a legacy approach or interface requirements when working with multiple DLT platforms? 
2. Would complete historical information on the original list of transfers be a requirement?
3. Also could you provide a JSON example of the type of transfers information breakdown you're looking highlighting the accounts, nodes amount summary for given transfers scenarios. This would help clarify things

In terms of the requests, i'm working on [PR 1823](https://github.com/hashgraph/hedera-mirror-node/pull/1823) which should address your request #4

**Comment by carlcherry** [comment]
Use Case 1 - our service pulls transactions out of the mirror node and imports them into our system.  We do this by keeping track of the index (height in our terms) that we last imported, check the mirror node to see if there are any new record files since we last checked.  You can thing of this as a "getheight" method where the returned index is the max index in record_file.

Use Case 2 - when we import records into our system we group them into a "psuedo" block: i.e. they all belong in the same record file. So we would need some REST api to get all transactions belonging to that record file.  Either by index or by consensus_start, consensus_end.

In our mirror nodes we already have complete history on the original transfers (I think). So all that is needed here in REST api changes.

I will work on the JSON samples and get back to you.

**Comment by carlcherry** [comment]
There are probably too many examples to consider - the simple approach would be to return both crypto_transfers (aggregated by account) AND non_fee_transfers (non-aggregated, controlled by a flag) in the API and I can figure things out in my code.

Currently I assume the following:
1. All nodes <= 99 are system nodes and those values are always assumed to be fees of some kind.
2. If the total for crypto_transfers and non_fee_transfers don't match or don't add up in case of multiple receivers then the difference must be a record fee.
3. If there are multiple senders, multiple receivers then I don't bother trying to match up sender/receiver.

**Comment by Nana-EC** [comment]
Hi @carl

**Comment by Nana-EC** [comment]
Hey @carlcherry 
So some notes on your assumptions
1. It should be all nodes <= 100
2. This is a fair assumption yes. In some earlier versions of the service there were items like threshold record fees that may have been paid by other accounts, not necessarily the main payer account. This is no longer the case but something to note for older transactions. Therefore all fees should be paid by the payer account which you can match with the transaction info.
3. This is understandable. The challenge is the aggregation is done prior to the transaction submission to the network. So there's no one that truly knows the transfer breakdown intent other than the submitter. In the simple cases it's easy to workout but in complex scenarios there may be a few ways to breakout the arithmetic and it's not 100% certain which way.

Planning to break out a few of your previous notes into tickets that we can address in coming sprints.

Thanks

**Comment by Nana-EC** [comment]
Hey @carlcherry 
I opened tickets that break out all your requests (referenced by this ticket). 

However, I'd like to know if your remaining greatest needs could be addressed by the following subset of enhancements. Note these are thoughts and suggestion and not final designs.
1. Exposing the `api/v1/records` endpoint solely as noted in #1933. This would give you the ability to pull the latest record file (`api/v1/records?order=desc&limit=1`) and get its timestamp range (`consensus_start`, `consensus_end`). You could then pull the given transaction from the transactions endpoint using `api/v1/transactions?timestamp=gte:<consensus_start>&timestamp=lte:<consensus_end>
2. Exposing `non_fee_transfer` in the transactions endpoint `api/v1/transactions`
   i. Adding a result (success|fail) column to the non_fee_transfer table to capture and distinguish failed transfers
   ii. Adding `includeNonFeeTransfers` flag to the transactions REST API endpoint as `api/v1/transactions?includeNonFeeTransfers=true`. The response would look like this with the additional "non_fee_transfer" object.

```json
{
  "transactions": [
    {
      "consensus_timestamp": 1234567890,
      "transaction_hash": "aGFzaA==",
      "valid_start_timestamp": 1234567890,
      "charged_tx_fee": 7,
      "memo_base64": null,
      "result": "SUCCESS",
      "name": "TRANSFER",
      "max_fee": 33,
      "valid_duration_seconds": 11,
      "node": "0.0.3",
      "transaction_id": "0.0.8-1234567890-000000006",
      "scheduled": false,
      "transfers": [
        {
          "account": "0.0.220479",
          "amount": 100000000
        },
        {
          "account": "0.0.219167",
          "amount": -100290430
        },
        {
          "account": "0.0.98",
          "amount": 270006
        },
        {
          "account": "0.0.3",
          "amount": 20424
        }
      ],
      "token_transfers": [
        {
          "token_id": "0.0.1234",
          "account": "0.0.220479",
          "amount": 1200
        },
        {
          "token_id": "0.0.1234",
          "account": "0.0.219167",
          "amount": -1200
        }
      ],
      "non_fee_transfers": [
        {
          "account": "0.0.220479",
          "amount": 100000000
        },
        {
          "account": "0.0.219167",
          "amount": -100000000
        }
      ]
    }
  ],
  "links": {
    "next": null
  }
}
```

**Comment by carlcherry** [comment]
I think that this would work for me.  We currently get the transaction
bytes as well but I don't think they are really that useful anyway, but
adding the raw bytes could also be an option.

**Comment by carlcherry** [comment]
One question - if a transaction fails then the only "transfers" should be (or could be) just fees?

**Comment by Nana-EC** [comment]
Sounds good. We have the tickets I opened so we'll have them scheduled for some future sprint. Let us know if there's urgency or priority on some of the request but we'll look to implement that base case for you.

In answer to your question. Yes in a failure transaction case the transfer would not go through but the fees would still be applicable.
See the following examples - http://testnet.mirrornode.hedera.com/api/v1/transactions?transactiontype=cryptotransfer&limit=5&result=fail

**Comment by Nana-EC** [comment]
> I think that this would work for me. We currently get the transaction bytes as well but I don't think they are really that useful anyway, but adding the raw bytes could also be an option.
> […](#)

Wanted to pick up on the minor comment here about `transaction_bytes`.
We added that per a previous request of yours. 
Remember to set the `hedera.mirror.importer.parser.record.entity.persist.transactionBytes` to true to ensure it's populated in the table.

Also when you say "raw bytes could also be an option" what are you referring to (want to make sure we're not mixing up jargon here. You should be able to pull everything you need from `transaction_bytes`.

In the above if what you meant by raw bytes was the signed transaction body bytes, see the logic we use to pull that out of the Transaction [here](https://github.com/hashgraph/hedera-mirror-node/blob/master/hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/domain/RecordItem.java#L109)
The `Transaction` object parsed in is just the result of `Transaction.parseFrom(transactionBytes);` from [here](https://github.com/hashgraph/hedera-mirror-node/blob/master/hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/domain/RecordItem.java#L72). 

Let me know if that clears things up

**Comment by Nana-EC** [comment]
Also I believe I've captured the action items based on the original ask in each of the noted tickets.
The base case as discussed to unblock you should be addressed by the implementation of [#1936](https://github.com/hashgraph/hedera-mirror-node/issues/1936)  and [#1979](https://github.com/hashgraph/hedera-mirror-node/issues/1979)

Based on this I'll be looking to close this ticket at the end of the sprint and utilize the smaller tickets, unless you thinks there's something else - such as the `transaction_bytes` discussion above.

Thanks

**Comment by Nana-EC** [comment]
Closing this based on above comments. If there's an issue not addressed we can cover in those related tickets.
Thanks
